_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+function.
+4. Lastly, we wrap everything in `FullTextQuery` and fetch result list by calling `getResultList()` function.
+
+One thing to note here is that although we are performing a query on Elasticsearch, Hibernate will still fire a query on the database to fetch the full entity. Which make sense, because as we saw in the previous section we didn't store all the fields of 
+`Post` entity and those fields still needed to be retrieved. If you only wanted to fetch what's stored in your index anyway and think this database call is redundant, I know just the way. Keep reading on to find out.
+
+Let's move to the next query. 
+
+Let's retrieve all the posts whose `likeCount` is greater than 1000 and should optionally contain 'food' hashtag and 
+'Literature' tag.
+
+```java
+public List<Post> getBasedOnLikeCountTags(Long likeCount, 
+                                          String hashTags, 
+                                          String tag){
+
+  FullTextEntityManager fullTextEntityManager = 
+                        Search.getFullTextEntityManager(entityManager);
+  QueryBuilder qb = fullTextEntityManager
+            .getSearchFactory()
+            .buildQueryBuilder()
+            .forEntity(Post.class)
+            .get();
+
+  Query likeCountGreater = qb.range()
+            .onField("likeCount")
+            .above(likeCount)
+            .createQuery();
+
+  Query hashTagsQuery = qb.keyword()
+            .onField("hashTags")
+            .matching(hashTags)
+            .createQuery();
+
+  Query tagQuery = qb.keyword()
+            .onField("tag")
+            .matching(tag)
+            .createQuery();
+
+  Query finalQuery = qb.bool()
+            .must(likeCountGreater)
+            .should(tagQuery)
+            .should(hashTagsQuery)
+            .createQuery();
+
+  FullTextQuery fullTextQuery = fullTextEntityManager
+                .createFullTextQuery(finalQuery, Post.class);
+  fullTextQuery.setSort(qb.sort().byScore().createSort());
+  return (List<Post>) fullTextQuery.getResultList();
+}
+```
+
+As you can see in the snippet above for `likeCount` we are using range query. Using only `above()` is equivalent to `>=` 
+operator. If you want to exclude the limits, just call `excludeLimit()` after `above()`.
+  
+For the other two fields, we have again used a keyword query. 
+
+Now, it's time to combine all queries. To do so, we will make use of `QueryBuilder` object's `bool()` function which 
+provides us with verbs such as `should()`, `must()` and `not()`. I have used `must()` for `likeCount` query and `should()`
+for the rest as they are optional. Optional queries wrapped in `should()` contribute to the relevance score. For instance document with tag matching, 'Literature' will score higher.
+
+A couple of paragraphs ago I had promised you that I will show you a way to avoid queries to the database. So, here it is, and 
+it's called **Projection**.
+```java
+public List<User> getUserByFirstWithProjection(String first, 
```

Comment:
By the way, this code formatting is not optimal for code snippets in blog posts because it uses up too much whitespace. I replaced the indentation with 2 spaces everywhere.